index.html

```diff
@@ -683,6 +683,284 @@ <h4>Proof Verification (ecdsa-2019)</h4>
+hashing algorithm</a> is defined in Section <a href="#hashing-jcs-ecdsa-2019"></a>,
+and the
+<a href="https://www.w3.org/TR/vc-data-integrity/#dfn-proof-serialization-algorithm">
+proof verification algorithm</a> is defined in Section
+<a href="#proof-verification-jcs-ecdsa-2019"></a>.
+          </p>
+        </section>
+
+        <section>
+          <h4>Transformation (jcs-ecdsa-2019)</h4>
+
+          <p>
+The following algorithm specifies how to transform an unsecured input document
+into a transformed document that is ready to be provided as input to the
+hashing algorithm in Section <a href="#hashing-jcs-ecdsa-2019"></a>.
+          </p>
+
+          <p>
+Required inputs to this algorithm are an
+<a data-cite="vc-data-integrity#dfn-unsecured-data-document">
+unsecured data document</a> (<var>unsecuredDocument</var>) and
+transformation options (<var>options</var>). The
+transformation options MUST contain a type identifier for the
+<a data-cite="vc-data-integrity#dfn-cryptosuite">
+cryptographic suite</a> (<var>type</var>) and a cryptosuite
+identifier (<var>cryptosuite</var>). A <em>transformed data document</em> is
+produced as output. Whenever this algorithm encodes strings, it MUST use UTF-8
+encoding.
+          </p>
+
+          <ol class="algorithm">
+            <li>
+If <var>options</var>.<var>type</var> is not set to the string
+`DataIntegrityProof` and <var>options</var>.<var>cryptosuite</var> is not
+set to the string `jcs-ecdsa-2019` then a `PROOF_TRANSFORMATION_ERROR` MUST be
+raised.
+            </li>
+            <li>
+Let <var>canonicalDocument</var> be the result of applying the
+JSON Canonicalization Scheme [[RFC8785]] to the <var>unsecuredDocument</var>.
+            </li>
+            <li>
+Set <var>output</var> to the value of <var>canonicalDocument</var>.
+            </li>
+            <li>
+Return <var>canonicalDocument</var> as the <em>transformed data document</em>.
+            </li>
+          </ol>
+        </section>
+
+        <section>
+          <h4>Hashing (jcs-ecdsa-2019)</h4>
+
+          <p>
+The following algorithm specifies how to cryptographically hash a
+<em>transformed data document</em> and <em>proof configuration</em>
+into cryptographic hash data that is ready to be provided as input to the
+algorithms in Section <a href="#proof-serialization-jcs-ecdsa-2019"></a> or
+Section <a href="#proof-verification-jcs-ecdsa-2019"></a>.
+          </p>
+
+          <p>
+The required inputs to this algorithm are a <em>transformed data document</em>
+(<var>transformedDocument</var>) and <em>canonical proof configuration</em>
+(<var>canonicalProofConfig</var>). A single <em>hash data</em> value represented as
+series of bytes is produced as output.
+          </p>
+
+          <ol class="algorithm">
+            <li>
+Let <var>transformedDocumentHash</var> be the result of applying the
+SHA-256 (SHA-2 with 256-bit output) cryptographic hashing algorithm [[RFC6234]]
```

Comment:
Hmm, is this a weakness if we're creating a P-384 signature but using SHA-2-256 to hash? Shouldn't we be using SHA-2-384 here?

Reply:
What weakness would be created by doing this? Do you mean, do we want to cover the case that SHA-256 is broken but SHA-384 is not?

To be clear, we're not talking about using SHA-256 as the signing hash (which would be an interoperability problem that we definitely shouldn't do).

How a message is constructed is separable from how it's signed. The message will be signed by ES384 (which uses P-384 and SHA-384), no matter what we do right here.

One possible reason to use SHA-384 here is to reduce the number of primitives an implementation needs -- but, at the same time, that implies that implementations won't be split across message construction and signing. Using SHA-384 to construct the message would actually *burden* some implementations that would otherwise not need to change. For example, if a signing implementation presently uses SHA-256 to construct the message to be signed and then passes it off to an HSM, it doesn't need to change if a P-384 key is used to do the signing.

I'd think this comes down to a question around why P-384 would be used instead of P-256 in the first place. I doubt it's driven by the use of SHA-384. It's just that SHA-384 must be used as the signing hash with P-384 for interoperability purposes. We should be careful not to needlessly apply "strength matching" principles.